### DISCUSSION TITLE
[Question] Is there a way to be able to import non standard libraries?

### DISCUSSION THREAD

**Discussion by nmacafee** [pybricks/support#264] [Q&A] [answered]
**Question**
Is there a way to import non standard libraries? Using Mindstorm Robot Inventory Tricky robot.

For example:
import time
from time import sleep

**Context**
We are doing a hack at work, and are attempting to use a python API our company has made and use it on Tricky.

We are just starting to explore pybricks. Ideally we would like to import our own API. Use it, but that also requires a streaming connection to Tricky. 

**Screenshots**
There is a saying that a picture is worth a 1000 words. Screenshots really help to identify and solve problems.

**Comment by XilefTech** [comment]
In my [CharlieOSX](https://github.com/XilefTech/charlieOSX) I also used some 3rd party libraries. I just went the route of downloading the source code of the lib and putting it into my project folder that is then uploaded to the hub with the rest of my code.
I guess that wouldn't be practical if you got a lot of dependencies but so far it worked fine for me.

**Reply by nmacafee** [reply]
We could do that, but we have 5 hours left in our hack. Maybe in the future. Thanks for suggestion.

**Reply by Freezehot** [reply]
> In my [CharlieOSX](https://github.com/XilefTech/charlieOSX) I also used some 3rd party libraries. I just went the route of downloading the source code of the lib and putting it into my project folder that is then uploaded to the hub with the rest of my code. I guess that wouldn't be practical if you got a lot of dependencies but so far it worked fine for me.

Could you maybe explain how you do/did this because I think thats what I want to do, but I dont know how.

**Reply by XilefTech** [reply]
sure
let's use picoweb as an example: 
the source code for it lies in it's repository: https://github.com/pfalcon/picoweb
so I just downloaded the /picoweb/ folder from the repository and pasted it next to my main.py file in the working directory of my current project. 
And that's it, we can now import picoweb in our program. Please note that this doesn't install any dependencies and you need to download them manually aswell. Most of the time, the dependencies are well documented somewhere within a given libraries documentation.

**Comment by dlech** [comment, marked as answer]
We have `from pybricks.tools import wait` instead of the standard `time.sleep()` function.

In the REPL, you can run `help('modules')` to see what modules are available. 

Related issues:
- https://github.com/pybricks/support/issues/236
- https://github.com/pybricks/support/issues/140

In general though, since we are using MicroPython instead of regular Python, using regular Python code is not really possible without significant modifications.

Although you could make some glue code like:

```python
def sleep(time):
    wait(time * 1000)
```

**Reply by nmacafee** [reply]
Ok thanks for the responses. I don't think without significant work, we'll be able hack our project today with pybricks. 

Our ultimate goal was to be able to remotely control tricky through our messaging cloud solution. Well, at least I learned more about pybricks for my own projects at home. Thanks!
